File: 8_String_To_Integer/string_to_integer.go
package string_to_integer

const (
	INT_MAX = 1<<31 - 1
	INT_MIN = (-1) << 31
)

func myAtoi(str string) int {
	data := []byte(str)
	sign := 1
	val := 0
	i := 0

	if len(data) == 0 {
		return 0
	}

	for i = 0; i < len(data) && data[i] == ' '; i++ {
	}

	if i < len(data) && (data[i] == '+' || data[i] == '-') {
		if data[i] == '+' {
			sign = 1
		} else {
			sign = -1
		}
		i++
	}

	for i < len(data) && (data[i] >= '0' && data[i] <= '9') {
		if val > INT_MAX/10 || (val == INT_MAX/10 && data[i]-'0' > 7) {
			if sign == 1 {
				return INT_MAX
			} else {
				return INT_MIN
			}
		}
		val = val*10 + int(data[i]-'0')
		i++
	}

	return val * sign
}

File: 8_String_To_Integer/string_to_integer.py
class Solution(object):
    def calcVal(self, str, i, sign):
        val = 0
        while i < len(str) and str[i].isdigit():
            v = str[i]
            val = val*10 + int(v)
            i = i + 1
        if sign == 1 and val > pow(2,31)-1:
            return pow(2,31) - 1
        if sign == -1 and val > pow(2,31):
            return sign*pow(2,31)
        return sign*val
            
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        if len(str) == 0:
            return 0
        v = str[0]
        val = 0
        sign = 1
        i = 0
        if v == ' ':
            while i + 1 < len(str) and v == ' ':
                i = i + 1
                v = str[i] 
        if v == '-':
            sign = -1
            i = i + 1
            v = str[i]
        if v == '+':
            sign = 1
            i = i + 1
            v = str[i]
        if v.isdigit():
            return self.calcVal(str, i, sign)
        return 0 

print Solution().myAtoi(" -42")
